raclients/src/frontend/proxy.rs

```diff
@@ -0,0 +1,114 @@
+// SPDX-License-Identifier: Apache-2.0
+
+use super::lib::{
+    alloc::{
+        string::{String, ToString},
+        vec,
+    },
+    core::{
+        fmt::{self, Display, Formatter},
+        num::TryFromIntError,
+    },
+    format, Vec,
+};
+
+#[derive(Debug)]
+pub enum Error {
+    Eof,
+    UnexpectedEof,
+    WriteZero,
+    Other(String),
+}
+
+impl Display for Error {
+    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
+        match self {
+            Self::Eof => write!(f, "End of file encountered"),
+            Self::UnexpectedEof => write!(f, "Unexpected end of file encountered"),
+            Self::WriteZero => write!(f, "Zero bytes written"),
+            Self::Other(s) => write!(f, "{}", &s),
+        }
+    }
+}
+
+pub trait RacProxyRead {
```

Comment:
I prefer the old implementation.

Write and Read are the common trait in std, so we tried to implement that trait just to read/write. So please add Write/Read trait as before this PR and the `RacProxyRead` that implement the proxy protocol, like the Proxy trait in the old version.

Honestly I don't understand the value of these changes, why the old version was not enough?

Reply:
I prefer the current implementation, esepcially w/r/t `{read, write}_vec`.

Most of the proxy handling is deserializing structs over a socket by reading and writing:

- length of bytes buffer
- buffer itself

`{read, write}_vec` does this without needing to duplicate any code.